fs/zipfs.py

```diff
@@ -4,24 +4,110 @@
 from __future__ import print_function
 from __future__ import unicode_literals
 
-from datetime import datetime
 import zipfile
+import stat
+
+from datetime import datetime
 
 import six
 
 from . import errors
 from .base import FS
 from .compress import write_zip
-from .enums import ResourceType
+from .enums import ResourceType, Seek
 from .info import Info
 from .iotools import RawWrapper
+from .permissions import Permissions
 from .memoryfs import MemoryFS
 from .opener import open_fs
 from .path import dirname, normpath, relpath
 from .time import datetime_to_epoch
 from .wrapfs import WrapFS
 
 
+class _ZipExtFile(RawWrapper):
+
+    def __init__(self, fs, name):
+        self._zip = _zip = fs._zip
+        self._end = _zip.getinfo(name).file_size
+        self._pos = 0
+        super(_ZipExtFile, self).__init__(_zip.open(name), 'r', name)
+
+    def read(self, size=-1):
+        if self._pos >= self._end:
+            return b''
+        elif size is None or size < 0:
+            size = self._end - self._pos
+            # NB(@althonos): do NOT replace by self._f.read() !
+            buf = b''.join([self._f.read(size-1), self._f._readbuffer[-1:]])
```

Comment:
For the special case of 2 strings, I'm pretty sure a string concatenation will be more efficient here and perhaps a little clearer...

```python
buf = self._f.read(size-1) + self._f._readbuffer[-1:]
```

Reply:
Is it ? 
I'm scared of doing string concatenation with `+` since a bad personal experience, but I'll follow you on that.